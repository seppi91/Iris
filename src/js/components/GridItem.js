import React from 'react';
import { connect } from 'react-redux';
import {
  uriType,
  scrollTo,
  sourceIcon,
} from '../util/helpers';
import Link from './Link';
import Icon from './Icon';
import Thumbnail from './Thumbnail';
import LinksSentence from './LinksSentence';
import { I18n } from '../locale';

class GridItem extends React.Component {
  componentDidMount() {
    const {
      mopidyActions,
      spotifyActions,
      spotifyAvailable,
      item,
    } = this.props;

    if (!item || item.images) return;

    switch (uriType(item.uri)) {
      case 'artist':
        if (spotifyActions && spotifyAvailable) {
          //spotifyActions.getArtistImages(item);
        }
        break;

      case 'album':
        // If Mopidy doesn't find any images, then it will pass on the call to LastFM
        if (mopidyActions) {
          mopidyActions.getImages([item.uri]);
        }
        break;

      default:
        break;
    }
  }

  shouldComponentUpdate = (nextProps) => {
    const { item } = this.props;
    return nextProps.item !== item;
  }

  onContextMenu = (e) => {
    const { onContextMenu } = this.props;
    if (onContextMenu) {
      onContextMenu(e);
    }
  }

  renderSecondary = ({
    uri,
    tracks_total = 0,
    tracks,
    followers,
    albums_uris,
    artists,
  }) => {
    switch (uriType(uri)) {
      case 'playlist':
        return (
          <span className="grid__item__secondary__content">
            <I18n path="specs.tracks" count={tracks.length || tracks_total} />
          </span>
        );

      case 'artist':
        return (
          <span className="grid__item__secondary__content">
            {followers && <I18n path="specs.followers" count={followers.toLocaleString()} />}
            {albums_uris && <I18n path="specs.albums" count={albums_uris.length} />}
          </span>
        );

      case 'album':
        return (
          <span className="grid__item__secondary__content">
            {artists && <LinksSentence nolinks items={artists} />}
          </span>
        );

      default:
        return (
          <span className="grid__item__secondary__content">
            {artists && <LinksSentence nolinks items={item.artists} /> }
            {followers && <I18n path="specs.followers" count={followers.toLocaleString()} />}
          </span>
        );
    }
  }

  render = () => {
    const {
      item: {
        album,
      } = {},
      link: customLink,
      type,
      show_source_icon,
      grid_glow_enabled,
    } = this.props;

    let { item } = this.props;

    if (!item) return null;
    if (album) item = { ...item, ...album };

    const link = customLink || `/${type}/${encodeURIComponent(item.uri)}`;

    return (
      <Link
        className={`grid__item grid__item--${type}`}
        to={link}
        onClick={scrollTo}
        onContextMenu={this.onContextMenu}
      >
        <Thumbnail
          glow={grid_glow_enabled}
          size="medium"
          className="grid__item__thumbnail"
          images={item.images || item.icons}
          type={type}
        />
        <div className="grid__item__name">
          {item.name ? item.name : <span className="opaque-text">{item.uri}</span>}
        </div>
        <div className="grid__item__secondary">
          {show_source_icon && (
            <Icon name={sourceIcon(item.uri)} type="fontawesome" className="source" />
          )}
          {this.renderSecondary(item)}
        </div>
      </Link>
    );
  }
}

const mapStateToProps = (state) => {
  const {
    ui: {
      grid_glow_enabled,
    },
  } = state;

  return {
    grid_glow_enabled,
  };
};

export {
  GridItem,
};

export default connect(mapStateToProps)(GridItem);
