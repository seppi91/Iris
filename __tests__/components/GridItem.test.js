import React from 'react';
import { BrowserRouter } from "react-router-dom";

// Testing-specific
import { shallow, mount, render } from 'enzyme';
const state = require('../state');

// Test subjects
import GridItem from '../../src/js/components/GridItem';

describe('<GridItem />', () => {

	it('should handle album', () => {
		var album = state.core.albums['jest:album:one'];
		var dom = shallow(<GridItem item={album} />);
		expect(dom.find('.grid__item__name').text()).toEqual('One');
		expect(dom.find('.grid__item__secondary__content').length).toBe(1);
	});

	it('should handle artist', () => {
		var artist = state.core.artists['jest:artist:alpha'];
		var dom = shallow(<GridItem item={artist} />);
		expect(dom.find('.grid__item__name').text()).toEqual('Alpha');
		expect(dom.find('.grid__item__secondary__content').childAt(0).render().text()).toEqual('123 followers');
		expect(dom.find('.grid__item__secondary__content').childAt(1).render().text()).toEqual('1 albums');
	});

	it('should handle playlist', () => {
		var playlist = state.core.playlists['jest:playlist:one'];
		var dom = shallow(<GridItem item={playlist} />);
		expect(dom.find('.grid__item__name').text()).toEqual('One');
		expect(dom.find('.grid__item__secondary__content').render().text()).toEqual('2 tracks');
	});
});