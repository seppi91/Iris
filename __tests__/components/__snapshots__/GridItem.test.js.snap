// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GridItem /> should handle album 1`] = `
<a
  className="grid__item grid__item--album "
  data-qa-file="Link"
  data-qa-node="RouterLink"
  href="/album/bG9jYWw6YWxidW06bWQ1OjY2ZmJlYTM1OTNiYTk2YTE1YTlkNDE4OGJlYmFiNTBi"
  onClick={[Function]}
  onContextMenu={[Function]}
>
  <div
    className="thumbnail thumbnail--loaded thumbnail--medium grid__item__thumbnail"
    data-qa-file="Thumbnail"
    data-qa-node="div"
  >
    <i
      className="icon icon--material thumbnail__placeholder"
      data-qa-file="Icon"
      data-qa-node="i"
    >
      album
    </i>
    <div
      className="thumbnail__image"
      data-qa-file="Thumbnail"
      data-qa-node="div"
      style={
        Object {
          "backgroundImage": "url(\\"/local/17338e740316f18dbb5e3331ac6be6c1-500x500.jpeg\\")",
        }
      }
    />
    <div
      className="thumbnail__actions"
      data-qa-file="Thumbnail"
      data-qa-node="div"
    />
  </div>
  <div
    className="grid__item__name"
    data-qa-file="GridItem"
    data-qa-node="div"
  >
    Sirens of the Sea
  </div>
  <div
    className="grid__item__secondary"
    data-qa-file="GridItem"
    data-qa-node="div"
  >
    <ul
      className="grid__item__secondary__content details"
      data-qa-file="GridItem"
      data-qa-node="ul"
    >
      <li
        data-qa-file="GridItem"
        data-qa-node="li"
      >
        <span
          className="undefined links-sentence"
          data-qa-file="LinksSentence"
          data-qa-node="span"
        >
          <span
            data-qa-file="LinksSentence"
            data-qa-node="span"
          >
            <span
              data-qa-file="LinksSentence"
              data-qa-node="span"
            >
              Above & Beyond
            </span>
          </span>
        </span>
      </li>
    </ul>
  </div>
</a>
`;

exports[`<GridItem /> should handle artist 1`] = `
<a
  className="grid__item grid__item--artist "
  data-qa-file="Link"
  data-qa-node="RouterLink"
  href="/artist/bG9jYWw6YXJ0aXN0Om1kNTo0ZjZlNGY5NzllMmM0MGM1ZTZhZDE3MzU4MDRjMjliYw=="
  onClick={[Function]}
  onContextMenu={[Function]}
>
  <div
    className="thumbnail thumbnail--loaded thumbnail--medium grid__item__thumbnail"
    data-qa-file="Thumbnail"
    data-qa-node="div"
  >
    <i
      className="icon icon--material thumbnail__placeholder"
      data-qa-file="Icon"
      data-qa-node="i"
    >
      perm_identity
    </i>
    <div
      className="thumbnail__image"
      data-qa-file="Thumbnail"
      data-qa-node="div"
      style={
        Object {
          "backgroundImage": "url(\\"https://i.scdn.co/image/4dc7080ef509c36203a131a0eab8dd5e4800d7c2\\")",
        }
      }
    />
    <div
      className="thumbnail__actions"
      data-qa-file="Thumbnail"
      data-qa-node="div"
    />
  </div>
  <div
    className="grid__item__name"
    data-qa-file="GridItem"
    data-qa-node="div"
  >
    Above & Beyond
  </div>
  <div
    className="grid__item__secondary"
    data-qa-file="GridItem"
    data-qa-node="div"
  >
    <ul
      className="grid__item__secondary__content details"
      data-qa-file="GridItem"
      data-qa-node="ul"
    >
      <li
        data-qa-file="GridItem"
        data-qa-node="li"
      >
        2 albums
      </li>
    </ul>
  </div>
</a>
`;

exports[`<GridItem /> should handle playlist 1`] = `
<a
  className="grid__item grid__item--playlist "
  data-qa-file="Link"
  data-qa-node="RouterLink"
  href="/playlist/bTN1OkxvY2FsJTIwdGVzdGVyLm0zdTg="
  onClick={[Function]}
  onContextMenu={[Function]}
>
  <div
    className="thumbnail thumbnail--loaded thumbnail--medium grid__item__thumbnail"
    data-qa-file="Thumbnail"
    data-qa-node="div"
  >
    <i
      className="icon icon--material thumbnail__placeholder"
      data-qa-file="Icon"
      data-qa-node="i"
    >
      queue_music
    </i>
    <div
      className="thumbnail__image"
      data-qa-file="Thumbnail"
      data-qa-node="div"
      style={
        Object {
          "backgroundImage": "url(\\"null\\")",
        }
      }
    />
    <div
      className="thumbnail__actions"
      data-qa-file="Thumbnail"
      data-qa-node="div"
    />
  </div>
  <div
    className="grid__item__name"
    data-qa-file="GridItem"
    data-qa-node="div"
  >
    Local tester
  </div>
  <div
    className="grid__item__secondary"
    data-qa-file="GridItem"
    data-qa-node="div"
  >
    <ul
      className="grid__item__secondary__content details"
      data-qa-file="GridItem"
      data-qa-node="ul"
    >
      <li
        data-qa-file="GridItem"
        data-qa-node="li"
      >
        13 tracks
      </li>
    </ul>
  </div>
</a>
`;
